src/H5ES.c

```diff
@@ -135,6 +184,41 @@ H5ESget_count(hid_t es_id, size_t *count /*out*/)
     FUNC_LEAVE_API(ret_value)
 } /* end H5ESget_count() */
 
+/*-------------------------------------------------------------------------
+ * Function:    H5ESget_time_estimate
+ *
+ * Purpose:     Retrieve an estimate for completing the operations in an event set
+ *
+ * Return:      SUCCEED / FAIL
+ *
+ * Programmer:	Quincey Koziol
+ *              Thursday, December 10, 2020
+ *
+ *-------------------------------------------------------------------------
+ */
+herr_t
+H5ESget_time_estimate(hid_t es_id, uint64_t *time_est /*out*/)
```

Comment:
I am also not a fan of this API call. I think it's going to be nigh impossible to make it useful.

Reply:
It's up to the VOL connector to figure it out, and it's a "quality of implementation" aspect of an asynchronous VOL connector.   Any connector that doesn't want to try to estimate operations can always return the "unknown estimate" value (UINT64_MAX) for a request and the event set will stop iterating over the requests in the event set and return the "unknown" value to the application.

I would have used the new "dynamic optional VOL operations" feature for it, but it requires the addition of a new VOL operation value for requests and there's no way to have the event set "know" about a dynamically registered operation.

LBNL is currently improving the async VOL connector to track the I/O statistics and return more meaningful time estimates, and we will work with applications to take advantage of this capability to balance their compute vs. I/O.  Having a time estimate for a set of I/O operations is a _big_ enabler for pushing applications to take advantage of async I/O - it's something they've never had the ability to predict before.   When an application retrieves the time estimate and it's a lot larger or smaller than their predicted compute time, they can adjust one or the other to better overlap and take more advantage of the system resources.

Reply:
Just spit-balling here but might there be a way here or in the future for a record of time an `unknown estimate` operation wound up taking and a place to store/hash that value so that if the situation is encountered again, instead of `unknown estimate` it can return that value based on previous experience? Thats likely well beyon what is necessary to get this initial implementation into the main lib but wanted to toss out the idea to possibly inform thinking.

Reply:
Good point - still up the VOL connector, but could be useful for that code.